core/src/main/scala/com/databricks/labs/remorph/parsers/snowflake/SnowflakeAstBuilder.scala

```diff
@@ -56,31 +56,130 @@ class SnowflakeAstBuilder extends SnowflakeParserBaseVisitor[TreeNode] {
     val tableName = ctx.object_name().id_(0).getText
-    NamedTable(tableName, Map.empty, is_streaming = false)
+    ir.NamedTable(tableName, Map.empty, is_streaming = false)
   }
+
+  override def visitTable_source_item_joined(ctx: Table_source_item_joinedContext): ir.Relation = {
+
+    def translateJoinType(joinType: SnowflakeParser.Join_typeContext): ir.JoinType = {
+      if (joinType == null || joinType.outer_join() == null) {
+        ir.InnerJoin
+      } else if (joinType.outer_join().LEFT() != null) {
+        ir.LeftOuterJoin
+      } else if (joinType.outer_join().RIGHT() != null) {
+        ir.RightOuterJoin
+      } else if (joinType.outer_join().FULL() != null) {
+        ir.FullOuterJoin
+      } else {
+        ir.UnspecifiedJoin
+      }
+    }
+
+    def buildJoin(left: ir.Relation, right: SnowflakeParser.Join_clauseContext): ir.Join = {
+
+      ir.Join(
+        left,
+        right.object_ref().accept(this),
+        None,
+        translateJoinType(right.join_type()),
+        Seq(),
+        ir.JoinDataType(is_left_struct = false, is_right_struct = false))
+    }
+    val left = ctx.object_ref().accept(this)
+    ctx.join_clause().asScala.foldLeft(left)(buildJoin)
+  }
+
 }
-class SnowflakeExpressionBuilder extends SnowflakeParserBaseVisitor[Expression] {
 
-  override def visitSelect_list_elem(ctx: SnowflakeParser.Select_list_elemContext): Expression = {
+class SnowflakeExpressionBuilder extends SnowflakeParserBaseVisitor[ir.Expression] {
+
+  override def visitSelect_list_elem(ctx: SnowflakeParser.Select_list_elemContext): ir.Expression = {
     val column = ctx.column_elem().accept(this)
     if (ctx.as_alias() != null) {
       ctx.as_alias().accept(this) match {
-        case Alias(_, name, metadata) => Alias(column, name, metadata)
+        case ir.Alias(_, name, metadata) => ir.Alias(column, name, metadata)
         case _ => null
       }
     } else {
       column
     }
   }
-  override def visitColumn_name(ctx: SnowflakeParser.Column_nameContext): Expression = {
-    Column(ctx.id_(0).getText)
+  override def visitColumn_name(ctx: SnowflakeParser.Column_nameContext): ir.Expression = {
+    ir.Column(ctx.id_(0).getText)
   }
 
-  override def visitAs_alias(ctx: SnowflakeParser.As_aliasContext): Expression = {
+  override def visitAs_alias(ctx: SnowflakeParser.As_aliasContext): ir.Expression = {
     val alias = ctx.alias().id_().getText
-    Alias(null, Seq(alias), None)
+    ir.Alias(null, Seq(alias), None)
+  }
+
+  override def visitPrimitive_expression(ctx: Primitive_expressionContext): Expression = {
+    val columnName = ctx.id_(0).getText
+    ir.Column(columnName)
   }
 }
+
+class SnowflakePredicateBuilder extends SnowflakeParserBaseVisitor[ir.Predicate] {
+  override def visitExpr(ctx: ExprContext): Predicate = {
+
+    def buildComparison(left: Expression, right: Expression, op: Comparison_operatorContext): Predicate = {
```

Comment:
if inner function is greater than 2-5 lines of code, let's make it a private method